fedimint-core/src/db/mod.rs

```diff
@@ -2165,12 +2165,87 @@ macro_rules! push_db_key_items {
     };
 }
 
+/// Context passed to the db migration _functions_ (pay attention to `Fn` in the
+/// name)
+///
+/// Typically should not be referred to directly, and instead by a type-alias,
+/// where the inner-context is set.
+///
+/// Notably it it has the (optional) module id (innacessible to the modules
+/// directly, but used internally) and an inner context `C` injected by the
+/// outer-layer.
+///
+/// `C` is generic, as in different layers / scopes (server vs client, etc.) a
+/// different (module-typed, type erased, server/client, etc.) contexts might be
+/// needed, while the database migration logic is kind of generic over that.
+pub struct DbMigrationFnContext<'tx, C> {
+    dbtx: DatabaseTransaction<'tx>,
+    module_instance_id: Option<ModuleInstanceId>,
+    ctx: C,
+}
+
+impl<'tx, C> DbMigrationFnContext<'tx, C> {
+    pub fn map<R>(self, f: impl FnOnce(C) -> R) -> DbMigrationFnContext<'tx, R> {
+        DbMigrationFnContext {
+            ctx: f(self.ctx),
+            dbtx: self.dbtx,
+            module_instance_id: self.module_instance_id,
+        }
+    }
+
+    pub fn split_dbtx_ctx<'s>(&'s mut self) -> (&'s mut DatabaseTransaction<'tx>, &'s C) {
+        let Self { dbtx, ctx, .. } = self;
+
+        (dbtx, ctx)
+    }
+    pub fn ctx(&self) -> &C {
+        &self.ctx
+    }
+
+    pub fn dbtx(&mut self) -> DatabaseTransaction {
+        if let Some(module_instance_id) = self.module_instance_id {
+            self.dbtx.to_ref_with_prefix_module_id(module_instance_id).0
+        } else {
+            self.dbtx.to_ref_nc()
+        }
+    }
+
+    // TODO make priv
+    pub fn module_instance_id(&self) -> Option<ModuleInstanceId> {
+        self.module_instance_id
+    }
+
+    #[doc(hidden)]
+    pub fn __global_dbtx(&mut self) -> &mut DatabaseTransaction<'tx> {
+        &mut self.dbtx
```

Comment:
Should this call `ensure_global` (or whatever that function is called)?

Reply:
I can actually delete it now, but I've added some more `ensure_global` around.